5.5/mysql/bin/run-mysqld.sh

```diff
@@ -1,43 +1,96 @@
 #!/bin/bash -e
 
+# Be paranoid and stricter than we should be.
+# https://dev.mysql.com/doc/refman/5.5/en/identifiers.html
+mysql_identifier_regex='^[a-zA-Z0-9_]+$'
+mysql_password_regex='^[a-zA-Z0-9_~!@#$%^&*()-=<>,.?;:|]+$'
+
 function usage {
 	echo "You must specify following environment variables:"
-	echo "  \$MYSQL_USER"
-	echo "  \$MYSQL_PASSWORD"
-	echo "  \$MYSQL_DATABASE"
+	echo "  \$MYSQL_USER (regex: '$mysql_identifier_regex')"
+	echo "  \$MYSQL_PASSWORD (regex: '$mysql_password_regex')"
+	echo "  \$MYSQL_DATABASE (regex: '$mysql_identifier_regex')"
+	echo "Optional:"
+	echo "  \$MYSQL_ROOT_PASSWORD (regex: '$mysql_password_regex')"
 	exit 1
 }
 
-test -z "$MYSQL_USER" && usage
-test -z "$MYSQL_PASSWORD" && usage
-test -z "$MYSQL_DATABASE" && usage
-
-if [ ! -d '/var/lib/mysql/data' ]; then
-
-	echo 'Running mysql_install_db ...'
-	scl enable mysql55 "mysql_install_db --user=mysql --datadir=/var/lib/mysql/data"
-	echo 'Finished mysql_install_db'
-
-	# These statements _must_ be on individual lines, and _must_ end with
-	# semicolons (no line breaks or comments are permitted).
-	# TODO proper SQL escaping on ALL the things D:
-	TEMP_FILE='/tmp/mysql-first-time.sql'
-	cat > "$TEMP_FILE" <<-EOSQL
-		DELETE FROM mysql.user ;
-		CREATE USER '${MYSQL_USER}'@'%' IDENTIFIED BY '${MYSQL_PASSWORD}' ;
-		GRANT ALL ON *.* TO '${MYSQL_USER}'@'%' WITH GRANT OPTION ;
-		DROP DATABASE IF EXISTS test ;
-		CREATE DATABASE IF NOT EXISTS \`${MYSQL_DATABASE}\` ;
-		FLUSH PRIVILEGES ;
-	EOSQL
+function valid_mysql_identifier {
+	local var="$1" ; shift
+	[[ "${var}" =~ $mysql_identifier_regex ]]
+}
 
-	set -- "$@" --init-file="$TEMP_FILE"
-fi
+function valid_mysql_password {
+	local var="$1" ; shift
+	[[ "${var}" =~ $mysql_password_regex ]]
+}
+
+valid_mysql_identifier "$MYSQL_USER"     || usage
+valid_mysql_password   "$MYSQL_PASSWORD" || usage
+valid_mysql_identifier "$MYSQL_DATABASE" || usage
 
-chown -R mysql:mysql /var/lib/mysql
+# Make sure env variables don't propagate to mysqld process.
+mysql_user="$MYSQL_USER" ; unset MYSQL_USER
```

Comment:
are you sure the env vars don't get propagated to mysql process this way? the lowercase vars will be the same....
i thought you are going to unset them when you starting the mysql server